### PR TITLE
Fixed #25452 -- improved EmailValidator according to various RFCs.

### PR DESCRIPTION
**Pull request by bak1an**
See https://code.djangoproject.com/ticket/25452 for the details.

The changes were made based on:

- http://tools.ietf.org/html/rfc3696#section-3 (Human readable email
  address validation guide)
- http://tools.ietf.org/html/rfc2822 (Internet Message Format, section 3.4.1)
- http://tools.ietf.org/html/rfc2821 (SMTP, section 4.5.3.1)
- http://tools.ietf.org/html/rfc1035 (Domain names, section 2.3.1)

Actual changes that were made:

- Added check to ensure that TLD does not start with '-' (actual #25452 issue)
- Allowed \040 (space) character to be in a quoted string local part
- Limited local part to be up to 64 characters, domain part up to 255 characters
- Added extra test cases to ensure that local part neither starts nor ends with
  '.' and does not have adjacent periods
- Allowed quoted '@', ' ' and '\' in local part (dot atom form)
- Added a few rather weird test cases from
  https://en.wikipedia.org/wiki/Email_address#Examples (no code changes have
  been made for this, it just worked out of the box)
- Added test cases to ensure that email can't start or end with a whitespace

### CODE REVIEW

**Comment by timgraham** [comment]
The main concern I here is to make sure it doesn't introduced any catastrophic backtracking problems. Have you thought about it?

**Comment by bak1an** [comment]
@timgraham I will check this next week and come back with my findings. Thanks!

**Comment by timgraham** [comment]
@shaib, could you lend regex expertise here? I think it would be worthwhile to restructure the regexes so we can add some comments as done here: https://github.com/django/django/blob/4f0cd0fd162122da96978b357ac9fc9534529410/django/core/validators.py#L89-L96

**Comment by nedbat** [comment]
I would recommend being very reluctant to add complexity to this regex, and to only do it where there are false negatives.

**Comment by MarkusH** [comment]
@collinanderson suggested in #django-dev, to go with HTML5's address validation: https://html.spec.whatwg.org/multipage/forms.html#valid-e-mail-address

I totally agree on that. The only way to really validate an address is sending an email and awaiting a confirmation code.

Since Django is mostly used for web applications and the form widget for an `EmailField` already uses `type="email"` which makes modern browsers to validate the input based on above mentioned regex, I'd suggest to follow the HTML5 regex and document this. Using a more and more complex regex will still yield false positives and false negatives. In the end, a spelling mistake may result in a technically valid address which belongs to a non-existing user account or domain.

**Comment by shaib** [comment]
I'm +0 on using htlm5 validation for email addresses; while largely simplifying, it is backwards-incompatible and I believe the idea needs to be discussed on the mailing list.

If we don't go that route, I think we should make sure `EmailValidator` and `URLValidator` use the same validation for domains (unless there's some other reason not to). As far as I understand, the current code allows unicode (non-ASCII) parts in the latter but not the former.

Also, @timgraham is correct about restructuring. I would rephrase the currently suggested code, for example, along the lines of
```#!python
    user_regex = _lazy_re_compile(
        r"("
        r"^{atom_char}+(\.{atom_char}+)*\Z"
        r"|"
        r'^"({quoted_str_char}|\\{escapable_quoted_str_char})*"\Z'
        r")"
        .format(atom_char=r"([-!#$%&'*+/=?^_`{}|~0-9A-Z]|\\[@ \\])",
                quoted_str_char=r"[\001-\010\013\014\016-\037\040!#-\[\]-\177]",
                escaped_quoted_str_car=r"[\001-\011\013\014\016-\177]"),
        re.IGNORECASE)
```

FWIW, as far as I understand, there shouldn't be opportunities for excessive backtracking on these expressions. Each of the branches in them is marked by delimiters.

**Comment by bak1an** [comment]
@shaib @timgraham @MarkusH 

How about rather minimal fix for #25452 ([like this one](https://github.com/django/django/compare/master...bak1an:ticket_25452_minimal)) while validation regex improvements are in discussion?

**Comment by shaib** [comment]
@bak1an I see very little gain in that. We're not very close to code freeze, and as noted above, making sure an email address is valid helps very little because it is very easy to provide valid yet wrong inputs. 
On top of that, either of the two ways forward currently considered will nullify it -- the HTML5 email regex will be imported as is, and `URLValidator`'s domain regex already excludes such invalid domains.

**Comment by timgraham** [comment]
The consensus on the mailing list seems to be to simplify the validation, not make it more comprehensive.
